src/Bundle/DependencyInjection/Driver/Doctrine/DoctrineORMDriver.php

```diff
@@ -44,14 +46,31 @@ protected function addRepository(ContainerBuilder $container, MetadataInterface
             $repositoryClass = $metadata->getClass('repository');
         }
 
+        $aliasId = $metadata->getServiceId('repository');
+        $repositoryFactoryDef = $container->getDefinition('sylius.doctrine.orm.container_repository_factory');
+        $managerReference = new Reference($metadata->getServiceId('manager'));
         $definition = new Definition($repositoryClass);
-        $definition->setArguments([
-            new Reference($metadata->getServiceId('manager')),
-            $this->getClassMetadataDefinition($metadata),
-        ]);
-        $definition->setPublic(true);
 
-        $container->setDefinition($metadata->getServiceId('repository'), $definition);
+        if ($repositoryClass === EntityRepository::class) {
+            $entityClass = $metadata->getClass('model');
+
+            $definition->setFactory([$managerReference, 'getRepository']);
+            $definition->setArguments([$entityClass]);
+            $definition->setPublic(true);
+
+            $container->setDefinition($aliasId, $definition);
+
+            $repositoryFactoryDef->addMethodCall('addGenericEntity', [$entityClass]);
+        } else {
+            // we cant use the doctrine factory as this will lead to an endless loop
+            $definition->setArguments([$managerReference, $this->getClassMetadataDefinition($metadata)]);
+            $definition->addTag(ServiceRepositoryCompilerPass::REPOSITORY_SERVICE_TAG);
+
+            $serviceAlias = new Alias($repositoryClass, true);
+
+            $container->setDefinition($repositoryClass, $definition);
+            $container->setAlias($aliasId, $serviceAlias);
```

Comment:
Would it remove the BC break?

```suggestion
            $definition->setArguments([$managerReference, $this->getClassMetadataDefinition($metadata)]);
            $container->setDefinition($aliasId, $definition);
            
            $doctrineDefinition = new Definition($repositoryClass);
            $doctrineDefinition->addTag(ServiceRepositoryCompilerPass::REPOSITORY_SERVICE_TAG);
            $doctrineDefinition->setFactory([new Reference('service_container', 'get']);
            $doctrineDefinition->setArguments([$aliasId]);
            $container->setDefinition($repositoryClass, $doctrineDefinition);
```

Reply:
adjusted accordingly, bc break is now gone